DPE: Enable ReflectionAdapter to signal that a property has changed

amzn-tmryan opened the pull request:
**Description**
Prior to this PR, the DPE didn't have a way of notifying editors that its internal document had changed (user changes a property, for example). This meant that editors such as the Asset Editor wouldn't know when they could save the file.

This PR makes generally the following changes to enable saving and other property change signalling from the ReflectionAdapter to interested parties (the AssetEditorWidget in this case):
- The ReflectionAdapter now has a PropertyChangeEvent which can signal to functions taking a PropertyChangeInfo struct as described above
- The ReflectionAdapter is already listening for property change events in it's message handler, so we call its NotifyPropertyChanged function from there with the relevant data
- The AzToolsFramework::AssetEditorWidget now subscribes to the ReflectionAdapter's PropertyChangeEvent, calling into the same execution paths that it did for the RPE from there

![dpe_enableDpeDirtyDomNotifications](https://user-images.githubusercontent.com/104796591/182969897-ace512a6-2cf3-47e3-accf-8bfac1a72ace.gif)

**Testing**
- Verified properties changed in the Asset Editor were retained after saving assets and reopening both the Asset Editor and the saved file
- Verified git tracked changed asset files during the above test (to make sure they were still saving to disk)

amzn-tmryan commented:
I have some outstanding questions about this approach. Spent a lot of time digging through the APIs, and looking for the right place to send events, etc. before landing here.

Questions:
- Should the DPE or its adapter (feels a little dirty either way) be using an `AzToolsFramework::IPropertyEditorNotify` similar to the RPE, or should the DPE be relying on events coming from its `DocumentAdapter`?
- Is what's in this PR sufficient? It was difficult finding a place that felt "right" to notify from in the DPE. I put it where it is because we've received a `ChangedEvent` from the adapter, and it contained a patch.

NicholasVanSickle commented:
I would advocate for handling this at the `ReflectionAdapter` level - we need to capture changes in reflection anyways to handle cases like undo, and a document change isn't necessarily a product of user input (though I suspect it will be in the majority of cases).

I'd be OK with a replacement to `IPropertyEditorNotify`, but lighter weight would likely be just adding an `AZ::Event` to the `ReflectionAdapter` for `OnPropertyChanged` - maybe a signature that takes a param struct like this, where we can pack more contextual information in as needed:
```
struct PropertyChangeInfo
{
    const AZ::Dom::Path& m_path;
    const AZ::Dom::Value& m_newValue;
    AZ::DocumentPropertyEditor::Nodes::ValueChangeType m_changeType; //to denote if it's a "finished" edit or an "in-progress" edit like a slider move - might be good to move that enum somewhere better too?
};
```

It might take some iteration to reach parity with what `IPropertyEditorNotify` does (e.g. I'm not sure offhand if we actually use `BeforePropertyModified`) but I think this would be a step in the right direction.

amzn-tmryan commented:
I've updated the PR with changes based on Nick's advice above. Updated the PR description also.

Still need to run this through some more testing, but I think the general idea is there. Let me know if there are improvements that we can make now or if there is something I missed. I'm uncertain if these changes properly enable any undo work, but hopefully this is a step in that direction.

I can think of a few things that could be fixed now or preferably in an upcoming PR(s):
- Reach parity with IPropertyEditorNotify, or at least handle all the use cases of the EntityPropertyEditor as a next step (I've seen at least one case of BeforePropertyModified being used)
- Integrate these changes with the EntityPropertyEditor and other editors that use the DPE

Also, I only moved the ValueChangeType scope out one namespace. There are other enums in there that seem like they could move with this one and I'm not quite familiar enough with the code yet to know if moving them into DocumentPropertyEditor will clutter the namespace, or if there is another spot for them.

---

Only sort of related to this PR, but is `static void InvokeChangeNotify(const AZ::Dom::Value& domNode)` static only because you needed to call it from the RpePropertyHandlerWrapper? Is there only ever one ReflectionAdapter or does it not matter which DOM/DPE a node is coming from? Curious because we may support multiple Entity Inspectors at some point.

NicholasVanSickle commented on Code/Framework/AzFramework/AzFramework/DocumentPropertyEditor/DocumentAdapter.h:
```diff
@@ -182,11 +199,15 @@ namespace AZ::DocumentPropertyEditor
         void NotifyContentsChanged(const Dom::Patch& patch);
+        //! Subclasses may call this to trigger a PropertyChangeEvent to notify the view that one of this adapter's
+        //! property editor instances has altered its value.
+        void NotifyPropertyChanged(const PropertyChangeInfo& changeInfo);
```
Hm I was thinking of having this notification specifically for `ReflectionAdapter` since it only really makes sense in that context - otherwise every adapter sort of implicitly has to implement this, when really it's property changes in the reflection system that we want to hook.

This might sound a little weird/pedantic, but it's a design thing: the aspiration is that business logic like "when this property changes, do X" is handled not by the view but by the adapter, to make complicated side effects like that encapsulated from the presentation layer and easier to test.

amzn-tmryan replied:
I don't think it's pedantic at all. My understanding of the DPE is still growing, so it's more of a misunderstanding on my part. I was thinking all DocumentAdapters were for property editing applications and not thinking of the ReflectionAdapter as being _the_ adapter for property editors (which makes sense because these properties are reflected). I see how the DocumentAdapter shouldn't necessarily need knowledge of the reflection anything (should have been a clue when all I ended up doing in DocumentAdapter::NotifyPropertyChanged is forwarding those event params along).

I'll get that changed!

NicholasVanSickle commented:
> Only sort of related to this PR, but is static void InvokeChangeNotify(const AZ::Dom::Value& domNode) static only because you needed to call it from the RpePropertyHandlerWrapper? Is there only ever one ReflectionAdapter or does it not matter which DOM/DPE a node is coming from? Curious because we may support multiple Entity Inspectors at some point.

This method is static because it's a helper that's used in a couple places (the actual property editor instance + container operations in `ReflectionAdapter` iirc) - there can and absolutely will be multiple `ReflectionAdapter` instances, but all of the state is contained in the `domNode` you're invoking on (that's how the change attribute itself is looked up)

NicholasVanSickle commented:
The implementation itself looks great, thank you! Apologies for asking for one last thing, but I'd like to advocate for moving the `NotifyPropertyChanged` logic down to `ReflectionAdapter` and having things connect directly to that instead of adding a signal to `DocumentPropertyEditor`. My hope is to event complicating the implementation of future adapters and to try to keep business logic strictly on the adapter level where possible.

Once it's moved I'm absolutely fine with keeping the interface as-is and doing any follow-up work for feature parity with `IPropertyEditorNotify` in follow-up PRs. If a quick synchronous sync would be beneficial, feel free to look me up on the O3DE discord!

amzn-tmryan commented:
Moved the PropertyChangeEvent, its params definition, and the change notification request function to ReflectionAdapter; this makes the DocumentAdapter domain agnostic once again.

Also removed the event handling from the AzToolsFramework::DocumentPropertyEditor so it can remain domain agnostic as well. The AssetEditorWidget now subscribes directly to the ReflectionAdapter's property change events.

NicholasVanSickle reviewed (approved):
Looks good to me! Nice and clean 😄

AMZN-daimini commented on Code/Framework/AzToolsFramework/AzToolsFramework/AssetEditor/AssetEditorWidget.h:
```diff
@@ -103,6 +103,9 @@ namespace AzToolsFramework
 
             void OnNewAsset();
 
+            // For subscribing to document property editor adater property specific changes
```
nit: adater -> adapter

amzn-tmryan replied:
I'll fix this real quick.

alexmontAmazon reviewed (commented):
great work!
